src/hotspot/share/opto/graphInvariants.cpp

```diff
@@ -0,0 +1,617 @@
+  AtSingleOutputOfType(bool (Node::*type_check)() const, const Pattern* pattern) : _type_check(type_check), _pattern(pattern) {
+  }
+
+  bool check(const Node* center, Node_List& steps, GrowableArray<int>& path, stringStream& ss) const override {
+    Node_List outputs_of_correct_type;
+    for (DUIterator_Fast imax, i = center->fast_outs(imax); i < imax; i++) {
+      Node* out = center->fast_out(i);
+      if ((out->*_type_check)()) {
+        outputs_of_correct_type.push(out);
+      }
+    }
+    if (outputs_of_correct_type.size() != 1) {
+      ss.print_cr("Non-unique output of expected type. Found: %d.", outputs_of_correct_type.size());
+      for (uint i = 0; i < outputs_of_correct_type.size(); ++i) {
+        outputs_of_correct_type.at(i)->dump("\n", false, &ss);
+      }
+      return false;
+    }
+    bool result = _pattern->check(outputs_of_correct_type.at(0), steps, path, ss);
+    if (!result) {
+      steps.push(outputs_of_correct_type.at(0));
+      path.push(LocalGraphInvariant::OutputStep);
+    }
+    return result;
+  }
+  bool (Node::*_type_check)() const;
+  const Pattern* const _pattern;
+};
+
+struct PatternBasedCheck : LocalGraphInvariant {
+  const Pattern* const _pattern;
+  explicit PatternBasedCheck(const Pattern* pattern) : _pattern(pattern) {}
+  CheckResult check(const Node* center, LazyReachableCFGNodes& reachable_cfg_nodes, Node_List& steps, GrowableArray<int>& path, stringStream& ss) const override {
+    return _pattern->check(center, steps, path, ss) ? CheckResult::VALID : CheckResult::FAILED;
+  }
+};
+
+struct IfProjections : PatternBasedCheck {
+  IfProjections()
+      : PatternBasedCheck(
+            And::make(
+                new HasNOutputs(2),
+                new AtSingleOutputOfType(&Node::is_IfTrue, new True()),
+                new AtSingleOutputOfType(&Node::is_IfFalse, new True()))) {
+  }
+  const char* name() const override {
+    return "IfProjections";
+  }
+  CheckResult check(const Node* center, LazyReachableCFGNodes& reachable_cfg_nodes, Node_List& steps, GrowableArray<int>& path, stringStream& ss) const override {
+    if (!center->is_If()) {
+      return CheckResult::NOT_APPLICABLE;
+    }
+    CheckResult r = PatternBasedCheck::check(center, reachable_cfg_nodes, steps, path, ss);
+    if (r == CheckResult::FAILED) {
+      if (reachable_cfg_nodes.is_node_dead(center)) {
+        // That's ok for dead nodes right now. It might be too expensive to collect for IGVN, but it will be removed in loop opts.
+        ss.reset();
+        return CheckResult::VALID;
+      }
+    }
+    return r;
+  }
+};
+
+struct PhiArity : PatternBasedCheck {
+  const Node* region_node = nullptr;
+  PhiArity()
+      : PatternBasedCheck(
+            And::make(
+                new HasAtLeastNInputs(1),
+                new AtInput(
+                    0,
+                    And::make(
+                        new NodeClass(&Node::is_Region),
+                        new Bind(region_node))))) {
```

Comment:
This sort of binding is kinda cool! Never thought of it before. Could be really cool for general pattern matching.
We would have to find a solution if there would be multiple bindings though ... I think that's not possible with your patterns, right? Is that a fundamental constraint?

Reply:
>  We would have to find a solution if there would be multiple bindings though ... I think that's not possible with your patterns, right? Is that a fundamental constraint?

Not sure what you mean? `And::make(new Bind(bla), AtInput(1, new Bind(bli)))`? You probably mean something else.


> If we could somehow already cast the Bind variable to Region. Could be tricky.
> Doing this is_Region and bind could be a very common idiom, so very useful.

Interesting... Not sure how with some template magic we don't have (like `Node::is<RegionNode>`) but probably doable with macros. I'll give it a try.

Reply:
> We would have to find a solution if there would be multiple bindings though ... I think that's not possible with your patterns, right? Is that a fundamental constraint?

Sorry, that was not very clear. Yes you can bind multiple variables already. But you cannot do a disjuction (or) with binding. That would be helpful if you wanted to match patterns like:
```
((x + a) + a)
or
(a + (x + a))
```
We do that sort of thing a lot in IGVN optimizations: we need to be prepared to iterate over all associative reorderings.